core-kotlin-modules/core-kotlin-collections-6/src/test/kotlin/com/baeldung/parallelOperationsCollections/ParallelOperationCollectionsUnitTest.kt

```diff
@@ -0,0 +1,205 @@
+
+    data class Person(val name: String, val age: Int, var isAdult: Boolean? = null)
+
+    private val people = listOf(
+        Person("Martin", 12),
+        Person("Ahmad", 42),
+        Person("Alina", 13),
+        Person("Alice", 30),
+        Person("Bob", 16),
+        Person("Charlie", 40)
+    )
+
+    private fun List<Person>.assertResultsTrue() {
+        assertThat(this).containsExactly(
+            Person("Bob", 16, false),
+            Person("Alice", 30, true),
+            Person("Charlie", 40, true),
+            Person("Ahmad", 42, true)
+        )
+    }
+
+    private fun Person.setAdult() {
+        this.isAdult = this.age >= 18
+        logger.info(this.toString())
+    }
+    
+    private fun Instant.printTotalTime() {
+        val totalTime = Duration.between(this, Instant.now()).toMillis()
+        logger.info("Total time taken: {} ms", totalTime)
+    }
+
+    @Test
+    fun `using coroutines for parallel operations`() = runBlocking {
+        logger.info("Using Coroutines")
+        val startTime = Instant.now()
+
+        val filteredPeople = people
+            .map { person ->
+                async {
+                    person.setAdult()
+                    person
+                }
+            }.awaitAll()
+            .filter { it.age > 15 }
+            .sortedBy { it.age }
+
+        startTime.printTotalTime()
+
+        filteredPeople.assertResultsTrue()
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    @Test
+    fun `using coroutines for parallel operations with Flow`() = runBlocking {
+        logger.info("Using Kotlin Flow")
+        val startTime = Instant.now()
+
+        val filteredPeople = people.asFlow()
+            .flatMapMerge { person ->
+                flow {
+                    emit(
+                        async {
+                            person.setAdult()
+                            person
+                        }.await()
+                    )
+                }
+            }
+            .filter { it.age > 15 }.toList()
+            .sortedBy { it.age }
+
+        startTime.printTotalTime()
+
+        filteredPeople.assertResultsTrue()
+    }
+
+    @Test
+    fun `using RxJava for parallel operations`() { // Observable.class from io.reactivex;
+        logger.info("Using RxJava")
+        val startTime = Instant.now()
+
+        val observable = Observable.fromIterable(people)
+            .flatMap(
+                {
+                    Observable.just(it)
+                        .subscribeOn(Schedulers.computation())
+                        .doOnNext { person -> person.setAdult() }
+                },
+                people.size // Uses maxConcurrency for the number of elements
+            )
+            .filter { it.age > 15 }
+            .toList()
+            .map { it.sortedBy { person -> person.age } }
+            .blockingGet()
+
+        startTime.printTotalTime()
+
+        observable.assertResultsTrue()
+    }
```

Comment:
And similarly, RxJava vs RxKotlin - we don't cover Observable.fromIterable(people) in the article, but it seems like quite an important difference

Reply:
Edit: I can see Observable.fromIterable now but what confused me is that the first code block we show is the RxKotlin one that uses people.toObservable

Reply:
CMIIW. The difference between RxJava and RxKotlin is only in how to convert Iterable to Observable.
RxJava uses public static <T> Observable<T> fromIterable(Iterable<? extends T> source)
while RxKotlin with the toObservable() extension function.
after that we will do the same code as we are in Kotlin project.